frame/contracts/src/chain_extension.rs

```diff
@@ -0,0 +1,393 @@
+use sp_std::{
+	marker::PhantomData,
+	vec::Vec,
+};
+
+pub use frame_system::Trait as SysTrait;
+pub use pallet_contracts_primitives::ReturnFlags;
+pub use sp_core::crypto::UncheckedFrom;
+pub use crate::exec::Ext;
+pub use state::Init as InitState;
+
+/// Result that returns a [`DispatchError`] on error.
+pub type Result<T> = sp_std::result::Result<T, DispatchError>;
+
+/// A trait used to extend the set of contract callable functions.
+///
+/// In order to create a custom chain extension this trait must be implemented and supplied
+/// to the pallet contracts configuration trait as the associated type of the same name.
+/// Consult the [module documentation](self) for a general explanation of chain extensions.
+pub trait ChainExtension {
+	/// Call the chain extension logic.
+	///
+	/// This is the only function that needs to be implemented in order to write a
+	/// chain extensions. It is called whenever a contract calls the `seal_call_chain_extension`
+	/// imported wasm function.
+	///
+	/// # Parameters
+	/// - `func_id`: The first argument to `seal_call_chain_extension`. Usually used to
+	///		determine which function to realize.
+	/// - `env`: Access to the remaining arguments and the execution environment.
+	///
+	/// # Return
+	///
+	/// In case of `Err` the contract execution is immediatly suspended and the passed error
+	/// is returned to the caller. Otherwise the value of [`RetVal`] determines the exit
+	/// behaviour.
+	fn call<E: Ext>(func_id: u32, env: Environment<E, InitState>) -> Result<RetVal>
+	where
+		<E::T as SysTrait>::AccountId: UncheckedFrom<<E::T as SysTrait>::Hash> + AsRef<[u8]>;
+
+	/// Determines whether chain extensions are enabled for this chain.
+	///
+	/// The default implementation returns `true`. Therefore it is not necessary to overwrite
+	/// this function when implementing a chain extension. In case of `false` the deployment of
+	/// a contract that references `seal_call_chain_extension` will be denied and calling this
+	/// function will return [`NoChainExtension`](Error::NoChainExtension) without first calling
+	/// into [`call`](Self::call).
+	fn enabled() -> bool {
+		true
+	}
+}
+
+/// Implementation that indicates that no chain extension is available.
+impl ChainExtension for () {
+	fn call<E: Ext>(_func_id: u32, mut _env: Environment<E, InitState>) -> Result<RetVal>
+	where
+		<E::T as SysTrait>::AccountId: UncheckedFrom<<E::T as SysTrait>::Hash> + AsRef<[u8]>,
+	{
+		// Never called since [`Self::enabled()`] is set to `false`. Because we want to
+		// avoid panics at all costs we supply a sensible error value here instead
+		// of an `unimplemented!`.
+		Err(Error::<E::T>::NoChainExtension.into())
+	}
+
+	fn enabled() -> bool {
+		false
+	}
+}
+
+/// Determines the exit behaviour and return value of a chain extension.
+pub enum RetVal {
+	/// The chain extensions returns the supplied value to its calling contract.
+	Converging(u32),
+	/// The control does **not** return to the calling contract.
```

Comment:
I' m a little curious about the design of `Converging(u32)`. What's the reason to return an `u32` to contract layar? The doc comment just say "returns the supplied value", but I wanna know more about this.
In following test file, I just see the return is `Ok(RetVal::Converging(func_id))`, return the func_id directly. Can you list more case to explain it? 
Thanks!

Reply:
Look at the return of the `seal_call_chain_extension`. It is exactly this value you can return here. It could be used to signal back information to the contract without forcing it to decode the copied back data. I imagine it will be used mostly for return codes.

Reply:
Get it. It looks like most time it would be a "C style" function.( input and output are in parameters, and return a code represents return codes)

Reply:
Yes it explicitly allows for using all parameters as primitives if needed. But I expect that most functions will need more complex types and hence need to use them as pointers to copy a scale encoded buffer.